nupic/embodied/utils/model_parts.py

```diff
@@ -0,0 +1,243 @@
+        # TODO: check that contiguous.view actually does what its supposed to
+        x = x.contiguous().view(
+            -1, self.flatten_dim
+        )  # dims is calculated manually, 84*84 -> 20*20 -> 9*9 ->7*7
+        # run through fully connected layer
+        x = self.fc(x)
+        if self.last_nonlinear is not None:
+            # apply activation function the the last layer
+            x = self.last_nonlinear(x)
+        if self.layernormalize:
+            # normalize the last layer
+            x = self.layernorm(x)
+        return x
+
+    def layernorm(self, x):
+        """Normalize a layer."""
+        m = torch.mean(x, -1, keepdim=True).to(self.device)
+        v = torch.std(x, -1, keepdim=True).to(self.device)
+        return (x - m) / (v + 1e-8)
+
+
+class small_deconvnet(torch.nn.Module):
+    def __init__(self, ob_space, feat_dim, nonlinear, ch, positional_bias, device):
+        super(small_deconvnet, self).__init__()
+        self.H = ob_space.shape[0]
+        self.W = ob_space.shape[1]
+        self.C = ob_space.shape[2]
+
+        self.feat_dim = feat_dim
+        self.nonlinear = nonlinear
+        self.ch = ch
+        self.positional_bias = positional_bias
+        self.device = device
+
+        self.sh = (64, 8, 8)
+        self.fc = torch.nn.Sequential(
+            torch.nn.Linear(feat_dim, np.prod(self.sh)), nonlinear()
+        ).to(self.device)
+
+        # The last kernel_size is 7 not 8 compare to the origin implementation,
+        # to make the output shape be [96, 96]
+        feat_list = [
+            [self.sh[0], 128, 4, (2, 2), (1, 1)],
+            [128, 64, 8, (2, 2), (3, 3)],
+            [64, ch, 7, (3, 3), (2, 2)],
+        ]
+        self.deconv = torch.nn.Sequential().to(self.device)
+        for i, f in enumerate(feat_list):
+            self.deconv.add_module(
+                "deconv_%i" % i,
+                torch.nn.ConvTranspose2d(
+                    f[0], f[1], kernel_size=f[2], stride=f[3], padding=f[4]
+                ),
+            )
+            if i != len(feat_list) - 1:
+                self.deconv.add_module("nl_%i" % i, nonlinear())
+
+        self.bias = (
+            torch.nn.Parameter(torch.zeros(self.ch, self.H, self.W), requires_grad=True)
+            if self.positional_bias
+            else None
+        )
+
+        self.init_weight()
+
+    def init_weight(self):
+        for m in self.fc:
+            if isinstance(m, torch.nn.Linear):
+                torch.nn.init.xavier_uniform_(m.weight.data)
+                torch.nn.init.constant_(m.bias.data, 0.0)
+        for m in self.deconv:
+            if isinstance(m, torch.nn.ConvTranspose2d):
+                torch.nn.init.xavier_uniform_(m.weight.data)
+                torch.nn.init.constant_(m.bias.data, 0.0)
+
+    def forward(self, features):
+        x = self.fc(features)
+        x = x.view((-1,) + self.sh)
+        x = self.deconv(x)
+        x = x[:, :, 6:-6, 6:-6]
```

Comment:
6:-6 seems very specific, what is it about? Can it be a function of some other number defined elsewhere in the class? I am always wary of hard-coded constants as they can bring some hidden bugs (as in some other constant is changed and then this becomes obsolete).